File: src/main/java/me/testaccount666/serversystem/commands/executables/moderation/AbstractModerationCommand.java
package me.testaccount666.serversystem.commands.executables.moderation;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand;
import me.testaccount666.serversystem.moderation.AbstractModeration;
import me.testaccount666.serversystem.moderation.AbstractModerationManager;
import me.testaccount666.serversystem.userdata.OfflineUser;
import me.testaccount666.serversystem.userdata.User;
import me.testaccount666.serversystem.utils.DurationParser;
import org.bukkit.command.Command;

import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static me.testaccount666.serversystem.utils.MessageBuilder.command;
import static me.testaccount666.serversystem.utils.MessageBuilder.general;

public abstract class AbstractModerationCommand extends AbstractServerSystemCommand {

    @Override
    public void execute(User commandSender, Command command, String label, String... arguments) {
        if (!checkBasePermission(commandSender, command)) return;
        if (arguments.length == 0) {
            general("InvalidArguments", commandSender).syntaxPath(getSyntaxPath(command)).label(label).build();
            return;
        }

        if (!isRemoveModeration(command) && arguments.length < 2) {
            general("InvalidArguments", commandSender).syntaxPath(getSyntaxPath(command)).label(label).build();
            return;
        }

        var targetOptional = ServerSystem.Instance.getUserManager().getUser(arguments[0]);
        if (targetOptional.isEmpty()) {
            general("PlayerNotFound", commandSender).target(arguments[0]).build();
            return;
        }
        var cachedUser = targetOptional.get();
        var targetUser = cachedUser.getOfflineUser();
        if (targetUser.getName().isEmpty()) {
            general("PlayerNotFound", commandSender).target(arguments[0]).build();
            return;
        }

        if (isRemoveModeration(command)) {
            handleRemoveModeration(command, commandSender, targetUser);
            return;
        }

        var duration = DurationParser.parseDuration(arguments[1]);
        if (duration == -2) {
            command("Moderation.InvalidDuration", commandSender).target(targetUser.getName().get()).build();
            return;
        }
        if (duration == 0) {
            command("Moderation.NotZero", commandSender).target(targetUser.getName().get()).build();
            return;
        }

        var currentTime = System.currentTimeMillis();
        var expireTime = duration == -1? -1 : currentTime + duration;

        var defaultReason = command("Moderation.DefaultReason", commandSender)
                .target(targetUser.getName().get()).prefix(false).send(false).build();
        if (defaultReason.isEmpty()) {
            ServerSystem.getLog().severe("(Command: ${command.getName()}) Default reason is empty! This should not happen!");
            general("ErrorOccurred", commandSender).label(label).build();
            return;
        }
        var reason = defaultReason.get();
        if (arguments.length > 2)
            reason = IntStream.range(2, arguments.length).mapToObj(index -> "${arguments[index]} ").collect(Collectors.joining());

        handleCreateModeration(command, commandSender, targetUser, expireTime, reason);
    }

    private void handleRemoveModeration(Command command, User commandSender, OfflineUser targetUser) {
        var moderationManager = getModerationManager(targetUser);
        var activeModeration = moderationManager.getActiveModeration();
        if (activeModeration.isEmpty()) {
            general("Moderation.${type(command)}.Remove.NoActiveModeration", commandSender).target(targetUser.getName().get()).build();
            return;
        }

        moderationManager.removeModeration(activeModeration.get());
        command("Moderation.${type(command)}.Remove.Success", commandSender).target(targetUser.getName().get()).build();
    }

    private void handleCreateModeration(Command command, User commandSender, OfflineUser targetUser, long expireTime, String reason) {
        var moderationManager = getModerationManager(targetUser);
        if (moderationManager.hasActiveModeration()) {
            general("Moderation.${type(command)}.Add.AlreadyActiveModeration", commandSender).target(targetUser.getName().get()).build();
            return;
        }

        var moderation = createModeration(command, commandSender, targetUser, expireTime, reason);
        moderationManager.addModeration(moderation);
        command("Moderation.${type(command)}.Add.Success", commandSender).target(targetUser.getName().get()).build();
        handlePostModeration(command, commandSender, targetUser, moderation);
    }

    protected abstract void handlePostModeration(Command command, User commandSender, OfflineUser targetUser, AbstractModeration moderation);

    protected abstract AbstractModeration createModeration(Command command, User commandSender, OfflineUser targetUser, long expireTime, String reason);

    protected abstract boolean checkBasePermission(User commandSender, Command command);

    protected boolean isRemoveModeration(Command command) {
        return command.getName().startsWith("un");
    }

    protected abstract AbstractModerationManager getModerationManager(OfflineUser targetUser);

    protected abstract String type(Command command);
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/moderation/AbstractModerationCommand.kt
package me.testaccount666.serversystem.commands.executables.moderation

import me.testaccount666.serversystem.ServerSystem.Companion.log
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand
import me.testaccount666.serversystem.extensions.*
import me.testaccount666.serversystem.moderation.AbstractModeration
import me.testaccount666.serversystem.moderation.AbstractModerationManager
import me.testaccount666.serversystem.userdata.*
import me.testaccount666.serversystem.utils.DurationParser
import org.bukkit.command.Command

abstract class AbstractModerationCommand<T : AbstractModeration> : AbstractServerSystemCommand() {
    override fun execute(commandSender: User, command: Command, label: String, vararg arguments: String) {
        val targetUser = getService<UserManager>().getUserOrNull(arguments[0]) ?: run {
            commandSender.generalMsg("PlayerNotFound") { target(arguments[0]) }
            return
        }
        val targetOfflineUser = targetUser.offlineUser.also {
            if (it.getNameOrNull() == null) {
                commandSender.generalMsg("PlayerNotFound") { target(arguments[0]) }
                return
            }
        }

        if (isRemoveModeration(command)) {
            handleRemoveModeration(command, commandSender, targetOfflineUser)
            return
        }

        val duration = DurationParser.parseDuration(arguments[1])
        if (duration == -2L) {
            commandSender.commandMsg("Moderation.InvalidDuration") { target(targetOfflineUser.nameSafe) }
            return
        }
        if (duration == 0L) {
            commandSender.commandMsg("Moderation.NotZero") { target(targetOfflineUser.nameSafe) }
            return
        }

        val currentTime = System.currentTimeMillis()
        val expireTime = if (duration == -1L) -1 else currentTime + duration

        val defaultReason = commandSender.commandMsg("Moderation.DefaultReason") {
            target(targetOfflineUser.nameSafe)
            prefix(false)
            send(false)
            blankError(true)
        }

        if (defaultReason.isEmpty()) {
            log.severe("(Command: ${command.name}) Default reason is empty! This should not happen!")
            commandSender.generalMsg("ErrorOccurred") { label(label) }
            return
        }
        var reason = defaultReason
        if (arguments.size > 2) reason = arguments.join(2)


        handleCreateModeration(command, commandSender, targetOfflineUser, expireTime, reason)
    }

    private fun handleRemoveModeration(command: Command, commandSender: User, targetUser: OfflineUser) {
        val moderationManager = getModerationManager(targetUser)
        val activeModeration = moderationManager.activeModeration ?: run {
            commandSender.commandMsg("Moderation.${type(command)}.Remove.NoActiveModeration") { target(targetUser.nameSafe) }
            return
        }

        moderationManager.removeModeration(activeModeration)
        commandSender.commandMsg("Moderation.${type(command)}.Remove.Success") { target(targetUser.nameSafe) }
        handlePostRemoveModeration(command, commandSender, targetUser)
    }

    private fun handleCreateModeration(command: Command, commandSender: User, targetUser: OfflineUser, expireTime: Long, reason: String) {
        val moderationManager = getModerationManager(targetUser)
        if (moderationManager.hasActiveModeration()) {
            commandSender.commandMsg("Moderation.${type(command)}.Add.AlreadyActiveModeration") { target(targetUser.nameSafe) }
            return
        }

        val moderation = createModeration(command, commandSender, targetUser, expireTime, reason)
        moderationManager.addModeration(moderation)
        commandSender.commandMsg("Moderation.${type(command)}.Add.Success") { target(targetUser.nameSafe) }
        handlePostModeration(command, commandSender, targetUser, moderation)
    }

    protected abstract fun handlePostRemoveModeration(command: Command, commandSender: User, targetUser: OfflineUser)

    protected abstract fun handlePostModeration(command: Command, commandSender: User, targetUser: OfflineUser, moderation: T)

    protected abstract fun createModeration(command: Command, commandSender: User, targetUser: OfflineUser, expireTime: Long, reason: String): T

    protected fun isRemoveModeration(command: Command) = command.name.startsWith("un")

    protected abstract fun getModerationManager(targetUser: OfflineUser): AbstractModerationManager<T>

    abstract fun type(command: Command?): String?
}
